### ISSUE TITLE
Prior fix seems to be missing from current version - The number of additional EC codewords must be 3 or more

### ISSUE THREAD

**Issue by ianbale** [metafloor/bwip-js#347]
Hi, 

I opened this issue back in January: [320](https://github.com/metafloor/bwip-js/issues/320 )

I just realised that I left my codebase using 3.0.1, since I got distracted between @terryburton saying he'd pushed a fix and waiting for 4.3.0 release. So I never managed to test that version. My apologies for that.

However, I have just tried it with @bwipjs/react-native@latest and I am seeing the same error. 

Any chance this fix can be applied to the current version?

The reason I need this is that the barcodes I am making are for UK rail tickets. The spec gives no room for any modification - eg. adding an extra layer to accommodate the data. So I need to be able to create my barcodes to the same spec as 3.0.1 did.

**Comment by metafloor** [comment]
Do you know what version of BWIPP contains the fixed code?  We are only about two months lagging (BWIPP latest is at 17-Aug and bwip-js is using the 18-Jun version), so surprised the fix from earlier this year is not incorporated.

**Comment by terryburton** [comment]
> I just realised that I left my codebase using 3.0.1, since I got distracted between @terryburton saying he'd pushed a fix and waiting for 4.3.0 release. So I never managed to test that version. My apologies for that.

The aforementioned fix is for the comment by @metafloor "do you see the errant qrcode reference in the raiseerror".

The lower limit for `ecaddchars` is retained since it is a sane requirement of the specification which is necessary to ensure that the message isn't corrected towards a misread.

I would only be prepared to get into the game of compromising the data security properties of a symbology by adding a dedicated encoder that named and shamed the relevant application, e.g. `rdg_wrongheaded_emulation_of_an_aztec_code`. We see all too many application standards and reference software written by so-called specialists that don't understand the first thing about the subject matter and then "optimise" the ISO standards to obtain smaller symbols — declaring victory as soon as the reader goes "beep". 

Before doing this, it seems fair that someone (not me!) should attempt to contact whoever maintains the application specification to tell them that they screwed up and are not compliant with ISO/IEC 24778, to give them the opportunity to correct things.

**Comment by ianbale** [comment]
@terryburton That does make me smile!

RDG are a clueless organisation that no longer employ anyone who understands anything about the systems they now manage. Even less so, anyone who actually understands technical details of any of the technology they rely upon. But sadly, they have been given control over all things tech related for rail travel within the UK.

The specification was written by Trainline.com (Largest UK / International rail ticket retailer) and adopted by RDG (who manage UK rail applications) many years ago. You'd be gobsmacked if I could share it (NDA)... But the real highlight is that the barcode contains "encrypted" data about the ticket. The problem is that if you do the encryption properly (ie. encrypt with public key and use private key in the scanner)  then the barcode payload is too large. So the spec says you have to use the keys in reverse. It took me ages to find an encryption library that would even permit me to do this!

Even when I pointed out how easy is is to reverse engineer the keys by doing the encryption backwards  especially as they always having some non-encrypted content identical to the first part of the encrypted content and other fixed items in the "encrypted" data. I even pointed them at a study where they have already broken "encryption" similar to what they are using.

As soon as someone has that key they can create fake tickets that are indistinguishable... They were still not willing to accept there is a problem or do anything about it!

The bottom line is that they had a certain amount of data they wanted encoded and it had to fit into a 7 layer barcode since an 8 layer one produced a code that regularly failed scanning (early mobile devices). Of course, screens on mobile device have improved and 8 layers is now fine to scan, but the spec has not (and will not) been updated.

To be fair, if they did they'd need to force dozens of companies to update apps, automated station gates, conductor scanning machines etc. It would cost a fortune and no one is going to be prepared to pay - at least not until someone actually gets caught making fake tickets and they can identify a very significant loss of revenue.

So, that leaves us developers either having to do these silly things with encryption/barcodes or not meeting the spec and not being permitted to issue tickets with our apps.

I'd very much appreciate a workaround so I can use the latest versions, but understand your position if you prefer not to. I can remain on v 3.0.1 for the Node.js app that makes the tickets. I'm just not sure if I can use that in my new React-Native app - a ticket designer.

**Comment by terryburton** [comment]
@ianbale What is the formal name of the specification?

**Comment by ianbale** [comment]
Hi @terryburton ,

The spec is designated RSPS3001.

When my ticking system was initially accredited, it was rejected because the barcode used 8 layers - which exceeded the limit of 7 layers. Hence why I ended up going back to 3.0.1 so I could stay within the permitted 7 layers and get the accreditation to go live.

However, whilst looking for the correct spec to quote to you, I discovered that the latest version (Published end of July - replacing the 2022 version) now says that the barcode should not exceed EIGHT layers...rather than 7 layers it used to specify. Interestingly, that change is not listed in the change log - hence why I overlooked it.

I suspect Trainline hit this problem and told RDG to change the spec. There is no way RDG worked this out. Sadly, RDG whilst RDG will listen to the big retailers who sell millions of tickets a month, they do not listen to smaller retailers - hence why my attempts to get this changed 4 years ago got nowhere...

Anyway, upshot is that I can now use an 8 layer barcode and so don't need to do anything special. So I am sorry to have taken up your time with this, but thank you for nudging me to re-check the spec. I very much appreciate the fact that you would have helped me out with this if the spec had not been updated and I was still limited to 7 layer barcodes.
